core-primitives/stf-executor/src/executor.rs

```diff
@@ -0,0 +1,440 @@
+		StfExecutor { ocall_api, state_handler }
+	}
+
+	/// Execute a trusted call on the STF
+	///
+	/// We distinguish between an error in the execution, which maps to `Err` and
+	/// an invalid trusted call, which results in `Ok(ExecutionStatus::Failure)`. The latter
+	/// can be used to remove the trusted call from a queue. In the former case we might keep the
+	/// trusted call and just re-try the operation.
+	fn execute_trusted_call_on_stf<PB, E>(
+		&self,
+		state: &mut E,
+		stf_call_signed: &TrustedCallSigned,
+		header: &PB::Header,
+		shard: &ShardIdentifier,
+		post_processing: StatePostProcessing,
+	) -> Result<ExecutedOperation>
+	where
+		PB: BlockT<Hash = H256>,
+		E: SgxExternalitiesTrait,
+	{
+		debug!("query mrenclave of self");
+		let mrenclave = self.ocall_api.get_mrenclave_of_self()?;
+		//debug!("MRENCLAVE of self is {}", mrenclave.m.to_base58());
+
+		let top_or_hash = top_or_hash::<H256>(stf_call_signed.clone(), true);
+
+		if let false = stf_call_signed.verify_signature(&mrenclave.m, &shard) {
+			error!("TrustedCallSigned: bad signature");
+			// do not panic here or users will be able to shoot workers dead by supplying a bad signature
+			return Ok(ExecutedOperation::failed(top_or_hash));
+		}
+
+		// Necessary because light client sync may not be up to date
+		// see issue #208
+		debug!("Update STF storage!");
+		let storage_hashes = Stf::get_storage_hashes_to_update(&stf_call_signed);
+		let update_map =
+			self.ocall_api.get_multiple_storages_verified(storage_hashes, header).map(into_map)?;
+
+		Stf::update_storage(state, &update_map.into());
+
+		debug!("execute STF");
+		let mut extrinsic_call_backs: Vec<OpaqueCall> = Vec::new();
+		if let Err(e) = Stf::execute(state, stf_call_signed.clone(), &mut extrinsic_call_backs) {
+			error!("Stf::execute failed: {:?}", e);
+			return Ok(ExecutedOperation::failed(top_or_hash));
+		}
+
+		let call_hash = blake2_256(&stf_call_signed.encode());
+		let operation = stf_call_signed.clone().into_trusted_operation(true);
+		let operation_hash = blake2_256(&operation.encode());
+		debug!("Operation hash {:?}", operation_hash);
+		debug!("Call hash {:?}", call_hash);
+
+		if let StatePostProcessing::Prune = post_processing {
+			state.prune_state_diff();
+		}
+
+		let execution_hashes =
+			ExecutionHashes::new(H256::from(call_hash), H256::from(operation_hash));
+
+		Ok(ExecutedOperation::success(execution_hashes, top_or_hash, extrinsic_call_backs))
+	}
+}
+
+impl<OCallApi, StateHandler> StfExecuteTrustedCall for StfExecutor<OCallApi, StateHandler>
+where
+	OCallApi: EnclaveAttestationOCallApi + EnclaveOnChainOCallApi + GetStorageVerified,
+	StateHandler: HandleState,
+{
+	fn execute_trusted_call<PB>(
+		&self,
+		calls: &mut Vec<OpaqueCall>,
+		stf_call_signed: &TrustedCallSigned,
+		header: &PB::Header,
+		shard: &ShardIdentifier,
+		post_processing: StatePostProcessing,
+	) -> Result<Option<(H256, H256)>>
```

Comment:
I quite dislike the fact that the result of this method is both in the returned value, and in the `&mut Vec<OpaqueCall>` parameter. In some parts, I replaced the `in/out` parameter by just a return value. But not here yet.